cluster-autoscaler/cloudprovider/gce/mig_info_provider.go

```diff
@@ -41,21 +46,95 @@ type MigInfoProvider interface {
 }
 
 type cachingMigInfoProvider struct {
-	mutex     sync.Mutex
-	cache     *GceCache
-	gceClient AutoscalingGceClient
-	projectId string
+	mutex                  sync.Mutex
+	cache                  *GceCache
+	gceClient              AutoscalingGceClient
+	projectId              string
+	concurrentGceRefreshes int
 }
 
 // NewCachingMigInfoProvider creates an instance of caching MigInfoProvider
-func NewCachingMigInfoProvider(cache *GceCache, gceClient AutoscalingGceClient, projectId string) MigInfoProvider {
+func NewCachingMigInfoProvider(cache *GceCache, gceClient AutoscalingGceClient, projectId string, concurrentGceRefreshes int) MigInfoProvider {
 	return &cachingMigInfoProvider{
-		cache:     cache,
-		gceClient: gceClient,
-		projectId: projectId,
+		cache:                  cache,
+		gceClient:              gceClient,
+		projectId:              projectId,
+		concurrentGceRefreshes: concurrentGceRefreshes,
 	}
 }
 
+// GetMigForInstance returns MIG ref for a given instance
+func (c *cachingMigInfoProvider) GetMigForInstance(instanceRef GceRef) (Mig, error) {
+	mig, found, err := c.getCachedMigForInstance(instanceRef)
+	if found {
+		return mig, err
+	}
+
+	foundCandidate, err := c.fillCandidateMigInstances(instanceRef)
+	if err != nil {
+		return nil, err
+	}
+
+	mig, found, err = c.getCachedMigForInstance(instanceRef)
+	if foundCandidate && !found {
+		c.cache.MarkInstanceMigUnknown(instanceRef)
+	}
+	return mig, err
+}
+
+func (c *cachingMigInfoProvider) getCachedMigForInstance(instanceRef GceRef) (Mig, bool, error) {
+	if migRef, found := c.cache.GetMigForInstance(instanceRef); found {
+		mig, found := c.cache.GetMig(migRef)
+		if !found {
+			return nil, true, fmt.Errorf("instance %v belongs to unregistered mig %v", instanceRef, migRef)
+		}
+		return mig, true, nil
+	} else if unknown := c.cache.IsMigUnknownForInstance(instanceRef); unknown {
+		return nil, true, nil
+	}
+	return nil, false, nil
+}
+
+// RefillAllMigInstances refills MIGs to instances mapping cache
+func (c *cachingMigInfoProvider) RefillAllMigInstances() error {
+	c.cache.InvalidateAllInstancesToMig()
+	migs := c.cache.GetMigs()
+	errors := make([]error, len(migs))
+	workqueue.ParallelizeUntil(context.Background(), c.concurrentGceRefreshes, len(migs), func(piece int) {
```

Comment:
Previous version would cancel upon receiving an error. Just to clarify - this change is by design, right? I'm assuming the intent here is to allow for better error handling?